File: blockchain/src/onchain/onchain.controller.ts
import {
  Controller,
  Get,
  Param,
  Post,
  Body,
  HttpStatus,
  HttpException,
} from '@nestjs/common';
import { OnchainService } from './onchain.service';
import { ResponseDTO } from '../dto/ResponseDTO';

@Controller('onchain')
export class OnchainController {
  constructor(private readonly onchainService: OnchainService) {}

  @Get('nft-count')
  async getNFTCount(): Promise<ResponseDTO<{ nftCount: string }>> {
    try {
      const nftCount = await this.onchainService.getNFTCount();
      return new ResponseDTO('NFT Count retrieved successfully', { nftCount });
    } catch (error) {
      throw new HttpException(
        new ResponseDTO('Failed to retrieve NFT Count', {
          error: error.message,
        }),
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Get('token-balance/:address')
  async getAddressBalance(
    @Param('address') address: string,
  ): Promise<ResponseDTO<{ balance: string }>> {
    try {
      const balance = await this.onchainService.getAddressBalance(address);
      return new ResponseDTO('Address Token balance retrieved successfully', {
        balance,
      });
    } catch (error) {
      throw new HttpException(
        new ResponseDTO('Failed to retrieve address token balance', {
          error: error.message,
        }),
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Post('transfer')
  async transfer(
    @Body('addressTo') addressTo: string,
    @Body('amount') amount: number,
    @Body('privateKey') privateKey: string,
  ): Promise<ResponseDTO<{ txResult: any }>> {
    try {
      const txResult = await this.onchainService.transfer(
        addressTo,
        amount,
        privateKey,
      );
      return new ResponseDTO('Transfer completed successfully', txResult);
    } catch (error) {
      throw new HttpException(
        new ResponseDTO('Failed to complete transfer', {
          error: error.message,
        }),
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Get('nft-balance/:address')
  async getAddressNFTBalance(
    @Param('address') address: string,
  ): Promise<ResponseDTO<{ balance: string }>> {
    try {
      const balance = await this.onchainService.getAddressNFTBalance(address);
      return new ResponseDTO('NFT balance retrieved successfully', { balance });
    } catch (error) {
      throw new HttpException(
        new ResponseDTO('Failed to retrieve NFT balance', {
          error: error.message,
        }),
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Get('nft-collection-name')
  async getNftCollectionName(): Promise<
    ResponseDTO<{ collectionName: string }>
  > {
    try {
      const collectionName = await this.onchainService.getNftCollectionName();
      return new ResponseDTO('NFT collection name retrieved successfully', {
        collectionName,
      });
    } catch (error) {
      throw new HttpException(
        new ResponseDTO('Failed to retrieve NFT collection name', {
          error: error.message,
        }),
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Get('nft-collection-owner')
  async getNftCollectionOwner(): Promise<ResponseDTO<{ owner: string }>> {
    try {
      const owner = await this.onchainService.getNftCollectionOwner();
      return new ResponseDTO('NFT collection owner retrieved successfully', {
        owner,
      });
    } catch (error) {
      throw new HttpException(
        new ResponseDTO('Failed to retrieve NFT collection owner', {
          error: error.message,
        }),
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Get('nft-owner/:nftId')
  async getNftOwner(
    @Param('nftId') nftId: number,
  ): Promise<ResponseDTO<{ owner: string }>> {
    try {
      const owner = await this.onchainService.getNftOwner(nftId);
      return new ResponseDTO('NFT owner retrieved successfully', { owner });
    } catch (error) {
      throw new HttpException(
        new ResponseDTO('Failed to retrieve NFT owner', {
          error: error.message,
        }),
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Get('token-uri/:nftId')
  async getTokenURI(
    @Param('nftId') nftId: number,
  ): Promise<ResponseDTO<{ tokenURI: string }>> {
    try {
      const tokenURI = await this.onchainService.getTokenURI(nftId);
      return new ResponseDTO('Token URI retrieved successfully', { tokenURI });
    } catch (error) {
      throw new HttpException(
        new ResponseDTO('Failed to retrieve token URI', {
          error: error.message,
        }),
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }
}

File: spring-boot-server/src/main/java/org/example/springbootserver/onchain/controller/OnchainController.java
package org.example.springbootserver.onchain.controller;

import lombok.RequiredArgsConstructor;
import org.example.springbootserver.global.dto.HttpResponseDTO;
import org.example.springbootserver.global.dto.HttpResponseDTOv2;
import org.example.springbootserver.onchain.service.OnchainService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/onchain")
@RequiredArgsConstructor
public class OnchainController {

    private final OnchainService onchainService;

    @GetMapping("/nft-count")
    public ResponseEntity<HttpResponseDTOv2> getNftCount() {
        return onchainService.getNftCount();
    }

    @GetMapping("/token-balance/{address}")
    public ResponseEntity<HttpResponseDTOv2> getTokenBalance(@PathVariable String address) {
        return onchainService.getTokenBalance(address);
    }

    @GetMapping("/token-balance")
    public ResponseEntity<HttpResponseDTOv2>  getTokenBalance() {
        return onchainService.getTokenBalance();
    }

    @GetMapping("/nft-balance/{address}")
    public ResponseEntity<HttpResponseDTOv2> getNftBalance(@PathVariable String address) {
        return onchainService.getNftBalance(address);
    }

    @GetMapping("/nft-collection-name")
    public ResponseEntity<HttpResponseDTOv2> getNftCollectionName() {
        return onchainService.getNftCollectionName();
    }

    @GetMapping("/nft-collection-owner")
    public ResponseEntity<HttpResponseDTOv2> getNftCollectionOwner() {
        return onchainService.getNftCollectionOwner();
    }

    @GetMapping("/nft-owner/{nftId}")
    public ResponseEntity<HttpResponseDTOv2> getNftOwner(@PathVariable String nftId) {
        return onchainService.getNftOwner(nftId);
    }

    @GetMapping("/token-uri/{nftId}")
    public ResponseEntity<HttpResponseDTOv2> getTokenUri(@PathVariable String nftId) {
        return onchainService.getTokenUri(nftId);
    }

}
